### PR TITLE
DRAFT: Exponential Smoothing

### PR DESCRIPTION
**Pull request by AdilZouitine**
This is a sketch I am currently implementing all the methods of the following [link](https://otexts.com/fpp2/expsmooth.html).

Then in a second step I will make it possible to use all the different methods in one class.

### CODE REVIEW

File: creme/time_series/exponential_smoothing.py [hunk 1]
```diff
@@ -1,56 +1,89 @@
...
     References:
         1. `Simple exponential smoothing <https://otexts.com/fpp2/ses.html>`_
     """
 
-    def __init__(self, alpha=0.5):
-        self.alpha = alpha
-        self.st = 0
+    def __init__(self, alpha=0.5, l0: float = 0):
+
+        if 0 <= alpha <= 1:
```
Comment (MaxHalford) [inline creme/time_series/exponential_smoothing.py]: I would instead do:

```python
if not 0 <= alpha <= 1:
    raise ValueError('blablabla')
self.alpha = alpha
```

That goes as well for the other parameters in the other classes :)

**Comment by AdilZouitine** [comment]
I'm currently at 8/9.
By the end of the weekend it'll be done.

File: creme/time_series/exponential_smoothing.py [hunk 2]
```diff
@@ -0,0 +1,566 @@
...
+        1. `Holt-Winters’ damped method <https://otexts.com/fpp2/holt-winters.html>`_
+    """
+
+    def __init__(self,
+                 m: int,
+                 alpha: float = 0.5,
+                 beta: float = 0.5,
+                 gamma: float = 0.5,
+                 phi: float = 0.5,
+                 s: Optional[List[float]] = None,
```
Comment (raphaelsty) [inline creme/time_series/exponential_smoothing.py]: Hi @AdilZouitine You should just import typing and set s: typing.Optional[typing.List[float]] = None :-)

File: creme/time_series/exponential_smoothing.py [hunk 3]
```diff
@@ -0,0 +1,566 @@
+from typing import List, Optional
+import collections  # replace by creme utils windows
+from . import base
+
+
+def _discount_sum(phi: float, h: int) -> list:
```
Comment (MaxHalford) [inline creme/time_series/exponential_smoothing.py]: Specify the type of the list. So I guess `typing.List[float]`?

File: creme/time_series/exponential_smoothing.py [hunk 4]
```diff
@@ -0,0 +1,566 @@
...
+    return sum([phi**i for i in range(1, h + 1)])
```
Comment (MaxHalford) [inline creme/time_series/exponential_smoothing.py]: Spacing around the `**` operator.

File: creme/time_series/exponential_smoothing.py [hunk 5]
```diff
@@ -0,0 +1,566 @@
...
+
+    Example:
+        ::
+            >>> #TODO
+
+    References:
+        1. `Simple exponential smoothing <https://otexts.com/fpp2/ses.html>`_
+    """
+
+    def __init__(self, alpha=0.5, l0: float = 0):
```
Comment (MaxHalford) [inline creme/time_series/exponential_smoothing.py]: Any reason why you type hinted one variable but not the other?

**Comment by AdilZouitine** [comment]
If you want check the formulas 
image

image

File: creme/time_series/exponential_smoothing.py [hunk 6]
```diff
@@ -0,0 +1,566 @@
...
+        self.alpha = alpha
+
+        self.lt = l0
+
+    def fit_one(self, y: float):
+        self.st = self.alpha * y + (1 - self.alpha) * self.lt
+
+        return self
+
+    def forecast(self, horizon: int) -> list:
```
Comment (MaxHalford) [inline creme/time_series/exponential_smoothing.py]: Again, type the list.

File: creme/time_series/exponential_smoothing.py [hunk 7]
```diff
@@ -0,0 +1,566 @@
...
+
+        self.m = m
+
+        self.alpha = alpha
+        self.beta = beta
+        self.gamma = gamma
+
+        # init s : [s_{t-m}, s_{t-m+1}, ..., s_{t}]
+        if s is None:
+            self.s = collections.deque([0 for _ in range(m)], maxlen=m)
```
Comment (MaxHalford) [inline creme/time_series/exponential_smoothing.py]: `[0] * m` also works

File: creme/time_series/exponential_smoothing.py [hunk 8]
```diff
@@ -0,0 +1,566 @@
...
+
+        self.m = m
+
+        self.alpha = alpha
+        self.beta = beta
+        self.gamma = gamma
+
+        # init s : [s_{t-m}, s_{t-m+1}, ..., s_{t}]
+        if s is None:
+            self.s = collections.deque([1 for _ in range(m)], maxlen=m)
```
Comment (MaxHalford) [inline creme/time_series/exponential_smoothing.py]: `[1] * m` also works

File: creme/time_series/exponential_smoothing.py [hunk 9]
```diff
@@ -0,0 +1,566 @@
...
+        self.lt = self.alpha * (y / st_m) + (1 - self.alpha) * (lt_1 + bt_1)
+        self.bt = self.beta * (self.lt - lt_1) + (1 - self.beta) * bt_1
+        st = self.gamma * (y / (lt_1 + bt_1)) + (1 - self.gamma) * st_m
+        self.s.append(st)
+
+        return self
+
+    def forecast(self, horizon: int) -> list:
+        return [(self.lt + h * self.bt) * self.s[h - self.m * ((
+            (h - 1) // self.m) + 1)] for h in range(1, horizon + 1)]
```
Comment (MaxHalford) [inline creme/time_series/exponential_smoothing.py]: Can you break this into multiple lines? For instance:

```py
return [
    (self.lt + h * self.bt) * self.s[h - self.m * (((h - 1) // self.m) + 1)]
    for h in range(1, horizon + 1)
]
```

In general I'm in favor of writing list comprehensions over multiple lines, I find that it increases readability. As you wish though! If you do change this, then please change all the other list comprehensions in your code :)

File: creme/time_series/exponential_smoothing.py [hunk 10]
```diff
@@ -0,0 +1,566 @@
...
+
+    Example:
+    ::
+        >>> #TODO
+
+    References:
+        1. `Holt-Winters’ additive method <https://otexts.com/fpp2/holt-winters.html>`_
+    """
+
+    def __init__(
```
Comment (MaxHalford) [inline creme/time_series/exponential_smoothing.py]: In `creme` we've mostly adopted a style of writing `__init__`'s in a single line, and line wrapping after 100 characters. Not really important though.

**Comment by lccambiaghi** [comment]
Hi, any update on this PR? I love your project and having ETS added to the pool of models would make it even more compelling :)

**Comment by AdilZouitine** [comment]
> Hi, any update on this PR? I love your project and having ETS added to the pool of models would make it even more compelling :)

Hey @lccambiaghi , I could move forward on Wednesday.
I'm missing a few tests, I think in a week or two I'll have it finished 😄

**Comment by AdilZouitine** [comment]
Sorry, I'm back after a few months on that PR.
Multiplicative methods have numerical instabilities and it doesn't seem a good choice to integrate them after reflection.

I think we should integrate a maximum of 6 methods (in my opinion 6 is a bit too much).
If it suits you, I'll finish the tests for the chosen methods (also some graphs) and the pr will be ready to be merged.

**Comment by MaxHalford** [comment]
Yep @AdilZouitine, sounds good :).

**Pull request closed without merging**